sample-schemas/demo_schema.erl

```diff
@@ -0,0 +1,57 @@
+-module(demo_schema).
+
+-include_lib("typerefl/include/types.hrl").
+
+-type ip() :: ipv4() | ipv6().
+-type ipv4() :: {0..255, 0..255, 0..255, 0..255}.
+-type ipv6() :: {0..65535, 0..65535, 0..65535, 0..65535, 0..65535, 0..65535, 0..65535, 0..65535}.
+
+-reflect_type([ip/0]).
+
+-export([fields/0]).
+
+fields() ->
+    [ {"foo.setting", fun setting/1}
+    , {"foo.endpoint", fun endpoint/1}
+    , {"foo.greet", fun greet/1}
+    ].
+
+setting(mapping) -> "app_foo.setting";
+setting(type) -> string();
+setting(converter) -> fun to_string/1;
+setting(_) -> undefined.
+
+endpoint(mapping) -> "app_foo.endpoint";
+endpoint(type) -> ip();
+endpoint(converter) -> fun to_ip/1;
```

Comment:
I think typerefl should support custom `(from|to)_string` callbacks for the types (so all IP addresses are parsed/printed in the same way). What do you think, @z8674558 @zmstone ?

Reply:
sounds nice. however, type conversion is not always `(from|to)_string`? e.g. when we deal with `percent` type, we would like to use `(from|to)_float`.

Reply:
Could you give an example of how conversion for percent type should work?

Reply:
I see. This situation can occur when HOCON parses value as a float? I think usually we will deal with the situations where some complex datatype (like IP or a comma-separated list) should be parsed from a string. In this case `hocon_schema` can use the default conversion, like you suggest here: https://github.com/emqx/hocon/pull/56#discussion_r618359516 , but that default conversion can benefit from inheriting the parser from the type.

Percent seems like a special case where a custom converter is required.

Reply:
ah I see. we can use custom converter in that situation.

Reply:
yes I agree that `(from|to)_string` is most commonly used.

Reply:
Sorry, took me a while to implement the prototype of the solution: https://github.com/k32/typerefl/pull/13

I forgot how painful Erlang metaprogramming is... :facepalm: